what_url/ada_adapter.py

```diff
@@ -0,0 +1,199 @@
+from what_url._ada_wrapper import ffi, lib
+
+URL_ATTRIBUTES = (
+    'href',
+    'username',
+    'password',
+    'protocol',
+    'port',
+    'hostname',
+    'host',
+    'pathname',
+    'search',
+    'hash',
+)
+PARSE_ATTRIBUTES = URL_ATTRIBUTES + ('origin',)
+
+
+def _get_str(x):
+    ret = ffi.string(x.data, x.length).decode('utf-8') if x.length else ''
+    return ret
+
+
+def check_url(s):
+    """
+    Returns ``True`` if *s* represents a valid URL, and ``False`` otherwise.
+
+    .. code-block:: python
+
+        >>> from what_url import check_url
+        >>> check_url('bogus')
+        False
+        >>> check_url('http://a/b/c/d;p?q')
+        True
+
+    """
+    try:
+        s_bytes = s.encode('utf-8')
+    except Exception:
+        return False
+
+    ada_url = lib.ada_parse(s_bytes, len(s_bytes))
+    try:
+        return lib.ada_is_valid(ada_url)
+    finally:
+        lib.ada_free(ada_url)
+
+
+def join_url(base_url, s):
```

Comment:
I think we should keep the API interface the same or similar to the WHATWG standard as possible to avoid confusion between different bindings.